File: Рудович/1654.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int minimumJumps(vector<int>& forbidden, int a, int b, int x) 
    {
        vector<int> d(4000001, -1);
        for (auto i : forbidden)
        {
            d[i] = INT_MAX;
        }
        d[0] = 0;
        queue<int> q, w;
        q.push(0);
        while(!q.empty() || !w.empty()){
            int t;
            if (w.empty())
            {
                t = q.front();
                q.pop();
            } 
            else if (q.empty())
            {
                t = w.front();
                w.pop();
            } 
            else if (d[w.front()] <= d[q.front()])
            {
                t = w.front();
                w.pop();
            }
            else
            {
                t = q.front();
                q.pop();
            }
            if (t + a < 4000001 && d[t + a] == -1)
            {
                if (t + a == x)
                {
                    return d[t] + 1;
                }
                d[t + a] = d[t] + 1;
                q.push(t + a);
            }
            if (t - b > 0 && d[t - b] == -1)
            {
                if (t - b == x)
                {
                    return d[t] + 1;
                }
                if (t - b + a < 4000001 && d[t - b + a] == -1)
                {
                    if (t - b + a == x)
                    {
                        return d[t] + 2;
                    }
                    d[t - b + a] = d[t] + 2;
                    w.push(t - b + a);
                }
            }
        }
        return d[x];
    }
};

File: Fedorova/1654.py
class Solution:
    def minimumJumps(self, forbidden: List[int], a: int, b: int, x: int) -> int:
        #BFS - Рассматриваем все возможные шаги вперед назад и запоминаем кол-во ходов
        #Во избежании повторов, также запоминаем посещенные позиции (+ шаг сделанный в позиции)
        limit = 2000 + a + b
        visited = set()
        queue = [(0,0,False)]
        while queue:
            new_position,res,step_b = queue.pop(0)
            if new_position in forbidden or (new_position,step_b) in visited:
                continue

            if new_position == x:
                return res

            if new_position + a < limit:
                queue.append((new_position + a,res + 1,False))

            if not step_b and new_position - b > 0:
                queue.append((new_position - b,res + 1,True))
            visited.add((new_position,step_b))
        return -1